src/common/intrusive_red_black_tree.h

```diff
@@ -0,0 +1,647 @@
+/*
+ * Copyright (c) 2018-2020 Atmosph�re-NX
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "common/parent_of_member.h"
+#include "common/tree.h"
+
+namespace Common {
+
+namespace impl {
+
+class IntrusiveRedBlackTreeImpl;
+
+}
+
+struct IntrusiveRedBlackTreeNode {
+
+private:
+    RB_ENTRY(IntrusiveRedBlackTreeNode) entry;
+
+    friend class impl::IntrusiveRedBlackTreeImpl;
+
+    template <class, class, class>
+    friend class IntrusiveRedBlackTree;
+
+public:
+    constexpr IntrusiveRedBlackTreeNode() : entry() { /* ... */
+    }
+};
+
+template <class T, class Traits, class Comparator>
+class IntrusiveRedBlackTree;
+
+namespace impl {
+
+class IntrusiveRedBlackTreeImpl {
+
+private:
+    template <class, class, class>
+    friend class ::Common::IntrusiveRedBlackTree;
+
+private:
+    RB_HEAD(IntrusiveRedBlackTreeRoot, IntrusiveRedBlackTreeNode);
+    using RootType = IntrusiveRedBlackTreeRoot;
+
+private:
+    IntrusiveRedBlackTreeRoot root;
+
+public:
+    template <bool Const>
+    class Iterator;
+
+    using value_type = IntrusiveRedBlackTreeNode;
+    using size_type = size_t;
+    using difference_type = ptrdiff_t;
+    using pointer = value_type*;
+    using const_pointer = const value_type*;
+    using reference = value_type&;
+    using const_reference = const value_type&;
+    using iterator = Iterator<false>;
+    using const_iterator = Iterator<true>;
+
+    template <bool Const>
+    class Iterator {
+    public:
+        using iterator_category = std::bidirectional_iterator_tag;
+        using value_type = typename IntrusiveRedBlackTreeImpl::value_type;
+        using difference_type = typename IntrusiveRedBlackTreeImpl::difference_type;
+        using pointer = typename std::conditional<Const, IntrusiveRedBlackTreeImpl::const_pointer,
+                                                  IntrusiveRedBlackTreeImpl::pointer>::type;
+        using reference =
```

Comment:
```suggestion
        using reference =
            std::conditional_t<Const, IntrusiveRedBlackTreeImpl::const_reference,
                                      IntrusiveRedBlackTreeImpl::reference>;
```

Ditto for `pointer` as well

Reply:
Causes build errors:
```
1>C:\Users\..\switch\yuzu\src\core\hle\kernel\k_address_arbiter.cpp(162,20): error C2819: type 'std::conditional<false,const Kernel::Thread *,Kernel::Thread *>' does not have an overloaded member 'operator ->'
1>C:\Users\..\switch\yuzu\src\common/intrusive_red_black_tree.h(314): message : see declaration of 'std::conditional<false,const Kernel::Thread *,Kernel::Thread *>'
1>C:\Users\..\switch\yuzu\src\core\hle\kernel\k_address_arbiter.cpp(162,55): error C2039: 'GetAddressArbiterKey': is not a member of 'Common::IntrusiveRedBlackTree<Derived,Common::IntrusiveRedBlackTreeMemberTraitsDeferredAssert<pointer-to-member(0x48),Derived>,Kernel::Thread::ConditionVariableComparator>::Iterator<false>'
```